parser-typechecker/src/Unison/Typechecker/Context.hs

```diff
@@ -867,75 +887,179 @@ noteTopLevelType e binding typ = case binding of
   _ -> btw $ topLevelComponent
     [(Var.reset (ABT.variable e), generalizeAndUnTypeVar typ, True)]
 
--- | Synthesize the type of the given term, updating the context in the process.
+synthesizeTop
+  :: Var v
+  => Ord loc
+  => Term v loc
+  -> M v loc (Type v loc)
+synthesizeTop tm = do
+  (ty, want) <- synthesize tm
+  ctx <- getContext
+  want <- substAndDefaultWanted want (out ctx)
+  when (not $ null want) . failWith $ do
+    AbilityCheckFailure
+      []
+      (Type.flattenEffects . snd =<< want)
+      ctx
+  applyM ty
+  where
+  out (Context es) = fmap fst es
+
+-- | Synthesize the type of the given term, updating the context in
+-- the process.  Also collect wanted abilities.
 -- | Figure 11 from the paper
-synthesize :: forall v loc . (Var v, Ord loc) => Term v loc -> M v loc (Type v loc)
-synthesize e | debugEnabled && traceShow ("synthesize"::String, e) False = undefined
+synthesize
+  :: Var v
+  => Ord loc
+  => Term v loc
+  -> M v loc (Type v loc, Wanted v loc)
+synthesize e | debugShow ("synthesize"::String, e) = undefined
 synthesize e = scope (InSynthesize e) $
   case minimize' e of
     Left es -> failWith (DuplicateDefinitions es)
     Right e -> do
-      Type.Effect'' es t <- go e
-      abilityCheck es
-      pure t
-  where
-  l = loc e
-  go :: (Var v, Ord loc) => Term v loc -> M v loc (Type v loc)
-  go (Term.Var' v) = getContext >>= \ctx -> case lookupAnn ctx v of -- Var
+      (Type.Effect'' es t, want) <- synthesizeWanted e
+      want <- coalesceWanted (fmap (Just e,) es) want
+      pure (t, want)
+
+wantRequest
+  :: Var v
+  => Ord loc
+  => Term v loc
+  -> Type v loc
+  -> (Type v loc, Wanted v loc)
+wantRequest loc ~(Type.Effect'' es t) = (t, fmap (Just loc,) es)
+
+synthesizeWanted
```

Comment:
Maybe a little comment here on the purpose of this function and a couple illustrative examples would be good.